### PR TITLE
Fix race condition in ADS

### PR DESCRIPTION
**Pull request by howardjohn**
So basically these two tests are doing this 3 second sleep for a full push to be triggered. This is bad from a testing standpoint (because we never want to sleep if possible) and leads to flakes - this test has been super flaky. Additionally, this is not really testing what we want - Envoy has no such 3 second sleep.

When removing this sleep, the test fails pretty often. What is happening here is a race between connection being added and a full push being triggered.

Normally we expect:

1. We send RDS request
2. RDS request completes
3. We add the connection
4. We send a full push to 1 connections

But because of a race we get:

1. We send RDS request
2. RDS request completes
3. We add the connection
4. We send a full push to 0 connections

This can be fixed by adding the connection before we try to respond to xDS requests. The added connection is not used anywhere in that code, so doing it first or last is not important.

This makes the problem much rarer, but there is still another race. When we send a response, we set the nonce for the response. However, we return from the send() function BEFORE we set the nonce asynchronously. This introduces a race where the next request comes in before we finish setting the nonce, and it is rejected. To fix this, we simply return from send only after we have set the nonce.

Fixes: https://github.com/istio/istio/issues/14504
Fixes: https://github.com/istio/istio/issues/11022
Fixes: https://github.com/istio/istio/issues/15822

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -224,6 +224,16 @@ func (s *DiscoveryServer) StreamAggregatedResources(stream ads.AggregatedDiscove
 				}
 			}
 
+			con.mu.Lock()
```
Comment (ramaraochavali) [inline pilot/pkg/proxy/envoy/v2/ads.go]: I know, you just moved the code from below, but what do you think about refactoring it as follows - I am just thinking taking a unconditional lock and releasing it based on conditions in if else might lead to subtle bugs if conditions change?

```
        bool added := false
	con.mu.Lock()
        added = con.added
        con.added = true
      con.mu.Unlock()
      if !added {
	s.addCon(con.ConID, con)
	defer s.removeCon(con.ConID, con)
      }
```

**Comment by hzxuzhonghu** [comment]
>This can be fixed by adding the connection before we try to respond to xDS requests. The added connection is not used anywhere in that code, so doing it first or last is not important.

@howardjohn This matters i think, if we add connection once a proxy connects, there may be full push triggered before the initial xDS requests. I am concerned whether this will introduce other races other than the redundant pushes.

**Comment by howardjohn** [comment]
@hzxuzhonghu I am pretty sure there will never be a full push before the request. First of all, even today we only guarantee the first request (CDS) completes, not all of them. But also, if you look at the `select` in this function, the XDS request will block the full push, so behavior should remain the same I think.

**Comment by hzxuzhonghu** [comment]
Could we add connection before the first rds pushes?

**Pull request merged**